src/line_parser.rs

```diff
@@ -0,0 +1,299 @@
+                _ => {
+                    panic!("Couldn't parse line string: {}", self.line_str());
+                }
+            }
+        }
+
+        match &self.state {
+            State::AwaitingNextToken => {
+                self.finish_note_if_sustaining(&mut line_notes);
+            }
+            _ => {
+                panic!(
+                    "Ended with half-parsed token {:?} when parsing line {}",
+                    self.state,
+                    self.line_str()
+                );
+            }
+        }
+
+        Line::new(line_notes)
+    }
+
+    fn forward(&mut self, num_chars: usize) {
+        self.next_index += num_chars;
+    }
+
+    fn get_note(&self, note_name: &str, octave: i8) -> Note {
+        match (note_name, octave) {
+            ("C", 3) => Note::C3,
+            ("Db", 3) => Note::Db3,
+            ("D", 3) => Note::D3,
+            ("Eb", 3) => Note::Eb3,
+            ("E", 3) => Note::E3,
+            ("F", 3) => Note::F3,
+            ("Gb", 3) => Note::Gb3,
+            ("G", 3) => Note::G3,
+            ("Ab", 3) => Note::Ab3,
+            ("A", 3) => Note::A3,
+            ("Bb", 3) => Note::Bb3,
+            ("B", 3) => Note::B3,
+            ("C", 4) => Note::C4,
+            ("Db", 4) => Note::Db4,
+            ("D", 4) => Note::D4,
+            ("Eb", 4) => Note::Eb4,
+            ("E", 4) => Note::E4,
+            ("F", 4) => Note::F4,
+            ("Gb", 4) => Note::Gb4,
+            ("G", 4) => Note::G4,
+            ("Ab", 4) => Note::Ab4,
+            ("A", 4) => Note::A4,
+            ("Bb", 4) => Note::Bb4,
+            ("B", 4) => Note::B4,
+            ("C", 5) => Note::C5,
+            ("Db", 5) => Note::Db5,
+            ("D", 5) => Note::D5,
+            ("Eb", 5) => Note::Eb5,
+            ("E", 5) => Note::E5,
+            ("F", 5) => Note::F5,
+            ("Gb", 5) => Note::Gb5,
+            ("G", 5) => Note::G5,
+            ("Ab", 5) => Note::Ab5,
+            ("A", 5) => Note::A5,
+            ("Bb", 5) => Note::Bb5,
+            ("B", 5) => Note::B5,
+            ("C", 6) => Note::C6,
+            ("Db", 6) => Note::Db6,
+            ("D", 6) => Note::D6,
+            ("Eb", 6) => Note::Eb6,
+            ("E", 6) => Note::E6,
+            ("F", 6) => Note::F6,
+            ("Gb", 6) => Note::Gb6,
+            ("G", 6) => Note::G6,
+            ("Ab", 6) => Note::Ab6,
+            ("A", 6) => Note::A6,
+            ("Bb", 6) => Note::Bb6,
+            ("B", 6) => Note::B6,
+            _ => {
+                panic!("Invalid note: {}{}", note_name, octave);
+            }
+        }
+    }
+
+    fn finished_parsing_beat(&mut self) {
+        self.current_beat_number = self.current_beat_number.add_sixteenths(1);
+        self.state = State::AwaitingNextToken;
+    }
+
+    fn mark_sustaining_note(&mut self, note_name: String, octave_char: char) {
+        let octave = octave_char.to_digit(10).unwrap() as i8;
+        self.currently_sustained_note =
```

Comment:
Seemed more straightforward/correct to store the information about the note currently being sustained in a separate field on `self` rather than eg pass around its info via whatever other states we might go through?